engine/language-server/src/main/java/org/enso/languageserver/boot/resource/LockedInitialization.java

```diff
@@ -0,0 +1,45 @@
+package org.enso.languageserver.boot.resource;
+
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.Executor;
+import java.util.concurrent.Semaphore;
+
+public abstract class LockedInitialization implements InitializationComponent {
+  private final Executor executor;
+  private final Semaphore lock = new Semaphore(1);
+  private boolean isInitialized;
+
+  protected LockedInitialization(Executor executor) {
+    this.executor = executor;
+    this.isInitialized = false;
+  }
+
+  @Override
+  public boolean isInitialized() {
+    return isInitialized;
+  }
+
+  public abstract void initComponent();
+
+  @Override
+  public CompletableFuture<Void> init() {
+    return CompletableFuture.runAsync(
+        () -> {
+          if (!isInitialized) {
+            try {
+              lock.acquire();
```

Comment:
We need to lock before checking (`!isInitialized`) to prevent double initialization

Reply:
Thanks for noticing, I must have omitted it during the cleanup in preparation of PR